src/box/space.c

```diff
@@ -550,9 +612,13 @@ space_apply_defaults(struct space *space, struct txn *txn,
 	const char *new_data_end = request->tuple_end;
 	size_t region_svp = region_used(&fiber()->gc);
 
-	bool changed = tuple_format_apply_defaults(space->format, &new_data,
-						   &new_data_end);
-	if (!changed) {
+	if (tuple_format_apply_defaults(space->format, &new_data,
+					&new_data_end) != 0) {
+		region_truncate(&fiber()->gc, region_svp);
+		return -1;
+	}
+	bool is_tuple_changed = new_data != request->tuple;
+	if (!is_tuple_changed) {
 		region_truncate(&fiber()->gc, region_svp);
```

Comment:
Nit: Better cleanup the region in `tuple_format_apply_defaults` on failure or in case the tuple is unchanged because the function doesn't return anything useful via the region in these cases. Up to you.

Reply:
Done.